index.js

```diff
@@ -101,50 +102,139 @@ function mapChars(domainName, useSTD3, processingOption) {
-
-
-function validateLabel(label, processingOption) {
+function validateLabel(label, processingOption, checkHyphens, checkBidi, checkJoiners) {
   if (label.substring(0, 4) === "xn--") {
-    label = punycode.decode(label.substring(4));
+    try {
+      label = punycode.decode(label.substring(4));
+    } catch (err) {
+      return {
+        label,
+        error: true
+      };
+    }
     processingOption = PROCESSING_OPTIONS.NONTRANSITIONAL;
   }
 
-  let error = false;
+  if (normalize(label) !== label) {
+    return {
+      label,
+      error: true
+    };
+  }
+
+  if (checkHyphens) {
+    if (label[2] === "-" && label[3] === "-" ||
+        label[0] === "-" || label[label.length - 1] === "-") {
+      return {
+        label,
+        error: true
+      };
+    }
+  }
 
-  if (normalize(label) !== label ||
-      (label[3] === "-" && label[4] === "-") ||
-      label[0] === "-" || label[label.length - 1] === "-" ||
-      label.includes(".") ||
-      label.search(combiningMarksRegex) === 0) {
-    error = true;
+  if (label.includes(".") ||
+      label.search(regexes.combiningMarks) === 0) {
+    return {
+      label,
+      error: true
+    };
   }
 
   const len = countSymbols(label);
   for (let i = 0; i < len; ++i) {
-    const status = findStatus(label.codePointAt(i));
+    const cp = label.codePointAt(i);
+    const status = findStatus(cp);
     if ((processingOption === PROCESSING_OPTIONS.TRANSITIONAL && status[1] !== "valid") ||
         (processingOption === PROCESSING_OPTIONS.NONTRANSITIONAL &&
          status[1] !== "valid" && status[1] !== "deviation")) {
-      error = true;
-      break;
+      return {
+        label,
+        error: true
+      };
+    }
+    if (cp > 0xFFFF) {
```

Comment:
Let's take the string `"a\uD800\uDC00b"`. When i = 1, codePointAt returns the code point that's > 0xFFFF. However, without the manual increment, when i = 2 codePointAt would return 0xDC00. That's just how codePointAt works.

We could, however, use for-of loop against the string which would handle this nuance for us.